src/renderer/features/network/NetworkSelector/model/network-selector-model.ts

```diff
@@ -59,62 +61,72 @@ sample({
 });
 
 sample({
-  clock: chainDisabled,
-  source: networkModel.$providers,
-  fn: (providers, chainId) => ({ chainId, providers }),
-  target: disconnectProviderFx,
+  clock: networkDisabled,
+  target: networkModel.events.chainDisconnected,
 });
 
 sample({
   clock: updateConnectionFx.doneData,
-  source: networkModel.$connections,
+  source: {
+    reconnectMap: $reconnectMap,
+    connections: networkModel.$connections,
+  },
   filter: (connection) => Boolean(connection),
-  fn: (connections, connection) => ({
-    ...connections,
-    [connection!.chainId]: connection,
+  fn: ({ reconnectMap, connections }, connection) => {
+    const chainId = connection!.chainId;
+    const update = { ...connections, [chainId]: connection };
+
+    const isOldEnabled = networkUtils.isEnabledConnection(connections[chainId]);
+    const isNewEnabled = networkUtils.isEnabledConnection(connection!);
+    if (isOldEnabled && isNewEnabled) {
+      return { update, reconnectMap: { ...reconnectMap, [chainId]: true } };
+    }
+
+    return { update };
+  },
+  target: spread({
+    reconnectMap: $reconnectMap,
+    update: networkModel.$connections,
   }),
-  target: networkModel.$connections,
 });
 
 sample({
   clock: updateConnectionFx.doneData,
-  source: networkModel.$providers,
-  filter: (_, connection) => {
-    return Boolean(connection) && networkUtils.isEnabledConnection(connection!);
-  },
-  fn: (providers, connection) => {
+  source: $reconnectMap,
+  filter: (_, connection) => Boolean(connection),
+  fn: (reconnectMap, connection) => {
     const chainId = connection!.chainId;
 
-    return providers[chainId] ? { reconnect: { chainId, providers } } : { start: chainId };
+    return reconnectMap[chainId] ? { disconnect: chainId } : { connect: chainId };
   },
   target: spread({
-    start: networkModel.events.chainConnected,
-    reconnect: reconnectProviderFx,
+    connect: networkModel.events.chainConnected,
+    disconnect: networkModel.events.chainDisconnected,
   }),
 });
 
 sample({
-  clock: [disconnectProviderFx.doneData, reconnectProviderFx.doneData],
-  source: networkModel.$providers,
-  fn: (providers, chainId) => {
-    const { [chainId]: _, ...rest } = providers;
-
-    return rest;
+  clock: networkModel.output.connectionStatusChanged,
```

Comment:
When connection status changed for a specific chain, we compare it with reconnect map and connect
Don't forget to remove the chain from the map